### PR TITLE
Replacing NotNullAttribute with thrown exceptions

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak \ @dougbu 

Preemptively checking this in to unbreak CI but I'll leave the PR open for comments.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/DefaultClientModelValidatorProvider.cs [hunk 1]
```diff
@@ -1,6 +1,10 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/DefaultClientModelValidatorProvider.cs]: Hmm, that's not right

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/DefaultClientModelValidatorProvider.cs]: ArgNullException is located under System

File: src/Microsoft.AspNet.Mvc.DataAnnotations/NumericClientModelValidator.cs [hunk 2]
```diff
@@ -16,11 +16,21 @@ public class NumericClientModelValidator : IClientModelValidator
         /// <inheritdoc />
         public IEnumerable<ModelClientValidationRule> GetClientValidationRules(ClientModelValidationContext context)
         {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/NumericClientModelValidator.cs]: Just curious: Is your compile module so smart that it nocied the `[NotNull` in the interface? Or did you catch this manually? I'm hoping the first of course...

File: src/Microsoft.AspNet.Mvc.DataAnnotations/DefaultClientModelValidatorProvider.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/DefaultClientModelValidatorProvider.cs]: Right but why is the duplicated license header also located under this `using`?

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/DefaultClientModelValidatorProvider.cs]: Ahh. Not sure why its doing that. Will fix.

File: src/Microsoft.AspNet.JsonPatch/Operations/OperationOfT.cs [hunk 3]
```diff
@@ -14,20 +13,48 @@ public Operation()
...
         {
+            if (op == null)
+            {
+                throw new ArgumentNullException(nameof(op));
+            }
+
+            if (path == null)
+            {
+                throw new ArgumentNullException(nameof(path));
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNet.JsonPatch/Operations/OperationOfT.cs]: Checks are done in the base constructor; same for the next constructor overload.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/DataTypeAttributeAdapter.cs [hunk 4]
```diff
@@ -17,9 +16,14 @@ public class DataTypeAttributeAdapter : DataAnnotationsClientModelValidator<Data
...
-            [NotNull] string ruleName)
+            string ruleName)
             : base(attribute)
         {
+            if (ruleName == null)
+            {
+                throw new ArgumentNullException(nameof(ruleName));
+            }
+
             if (string.IsNullOrEmpty(ruleName))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/DataTypeAttributeAdapter.cs]: Do we want a different exception type and message for `null` and `string.Empty`? Elsewhere we seem to _just_ check `string.IsNullOrEmpty()`.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/NumericClientModelValidator.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/NumericClientModelValidator.cs]: It looks up implemented interfaces and overriden methods.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/DataTypeAttributeAdapter.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/DataTypeAttributeAdapter.cs]: I'll remove the null check.

**Pull request closed without merging**